osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs

```diff
@@ -0,0 +1,138 @@
+                {
+                    CacheHttpClient = true,
+                    HttpClientCacheSize = 32,
+                    RegionEndpoint = RegionEndpoint.USWest1,
+                    UseHttp = true,
+                    ForcePathStyle = true,
+                    RetryMode = RequestRetryMode.Legacy,
+                    MaxErrorRetry = 5,
+                    Timeout = TimeSpan.FromMinutes(1),
+                });
+        }
+
+        public async Task StoreBeatmapSetAsync(uint beatmapSetId, byte[] beatmapPackage)
+        {
+            var stream = new MemoryStream(beatmapPackage);
+            using var archiveReader = new ZipArchiveReader(stream);
+
+            var files = new List<byte[]>();
+
+            foreach (string filename in archiveReader.Filenames)
+                files.Add(await archiveReader.GetStream(filename).ReadAllBytesToArrayAsync());
+
+            await Task.WhenAll(
+                uploadBeatmapPackage(beatmapSetId, beatmapPackage, stream),
+                uploadBeatmapFiles(files));
+        }
+
+        private Task<PutObjectResponse> uploadBeatmapPackage(uint beatmapSetId, byte[] beatmapPackage, MemoryStream stream)
+        {
+            return client.PutObjectAsync(new PutObjectRequest
+            {
+                BucketName = AppSettings.S3BucketName,
+                Key = getPathToPackage(beatmapSetId),
+                Headers =
+                {
+                    ContentLength = beatmapPackage.Length,
+                    ContentType = "application/x-osu-beatmap-archive",
+                },
+                InputStream = stream,
+            });
+        }
+
+        private Task uploadBeatmapFiles(List<byte[]> files)
+        {
+            return Parallel.ForEachAsync(
+                files,
+                async (file, cancellationToken) =>
+                {
+                    var fileStream = new MemoryStream(file);
+                    long length = file.Length;
+                    string sha2 = fileStream.ComputeSHA2Hash();
+
+                    await client.PutObjectAsync(new PutObjectRequest
+                    {
+                        BucketName = AppSettings.S3BucketName,
+                        Key = getPathToVersionedFile(sha2),
+                        Headers =
+                        {
+                            ContentLength = length,
+                        },
+                        InputStream = fileStream,
+                    }, cancellationToken);
+                });
+        }
+
+        public async Task ExtractBeatmapSetAsync(uint beatmapSetId, string targetDirectory)
+        {
+            using var response = await client.GetObjectAsync(AppSettings.S3BucketName, getPathToPackage(beatmapSetId));
+            // S3-provided `HashStream` does not support seeking which `ZipArchiveReader` does not like.
+            var memoryStream = new MemoryStream();
+            await response.ResponseStream.CopyToAsync(memoryStream);
```

Comment:
Very minor, but I think this change will avoid one memory copy (`CopyToAsync` creates a temporary buffer before copying the content across, doesn't have a `MemoryStream` specific shortcut method):

```diff
diff --git a/osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs b/osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs
index 1c81025..1839664 100644
--- a/osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs
+++ b/osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs
@@ -95,8 +95,7 @@ namespace osu.Server.BeatmapSubmission.Services
         {
             using var response = await client.GetObjectAsync(AppSettings.S3BucketName, getPathToPackage(beatmapSetId));
             // S3-provided `HashStream` does not support seeking which `ZipArchiveReader` does not like.
-            var memoryStream = new MemoryStream();
-            await response.ResponseStream.CopyToAsync(memoryStream);
+            using var memoryStream = new MemoryStream(await response.ResponseStream.ReadAllBytesToArrayAsync());
 
             using var archiveReader = new ZipArchiveReader(memoryStream);
 

```

Reply:
Needed to be `ReadAllRemainingBytesToArrayAsync()` because the non-remaining variant requires seekable streams, but applied in https://github.com/ppy/osu-server-beatmap-submission/pull/13/commits/71efcb0964f86c6ca381652fc32f5669aaa7de28.